include/envoy/network/transport_socket.h

```diff
@@ -0,0 +1,118 @@
+#pragma once
+
+#include "envoy/buffer/buffer.h"
+#include "envoy/common/pure.h"
+#include "envoy/network/connection.h"
+
+namespace Envoy {
+namespace Network {
+
+/**
+ * Action that should occur on a connection after I/O.
+ */
+enum class PostIoAction {
+  // Close the connection.
+  Close,
+  // Keep the connection open.
+  KeepOpen
+};
+
+/**
+ * Result of each I/O event.
+ */
+struct IoResult {
+  PostIoAction action_;
+
+  /**
+   * Number of bytes processed by the I/O event.
+   */
+  uint64_t bytes_processed_;
+};
+
+/**
+ * Callbacks used by transport socket instances to communicate with connection
+ */
+class TransportSocketCallbacks {
+public:
+  virtual ~TransportSocketCallbacks() {}
+
+  /**
+   * @return the file descriptor associated with the connection.
+   */
+  virtual int fd() PURE;
+
+  /**
+   * @return the connection interface.
+   */
+  virtual Connection& connection() PURE;
+
+  /**
+   * @return return whether the read buffer should be drained
```

Comment:
A bit more context on this would be nice now that it has moved away from where the code is using it.
You could maybe mention the example that this can be used to enforce yielding for configured read limits or some such

Reply:
Done